File: JS/Medium/Stacks/minStack.js
//155
var MinStack = function () {
    this.stack = [];
};

/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function (val) {
    if (this.stack.length == 0) {
        this.stack.push([val, val]);
    }
    else {
        //let minVal = this.stack.getMin();
        let minVal = this.stack[this.stack.length - 1][1];
        this.stack.push([val, Math.min(val, minVal)]);
    }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function () {
    this.stack.pop();
};

/**
 * @return {number}
 */
MinStack.prototype.top = function () {
    return this.stack[this.stack.length - 1][0];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function () {
    return this.stack[this.stack.length - 1][1];
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: C#/DSA/Medium/Stacks/MinStack.cs
//155
public class MinStack
{

    private Stack<List<int>> stack;
    public MinStack()
    {
        stack = new();
    }

    public void Push(int val)
    {
        if (stack.Count == 0)
        {
            stack.Push([val, val]);
        }
        else
        {
            var minVal = stack.Peek()[1];
            stack.Push([val, Math.Min(val, minVal)]);
        }
    }

    public void Pop()
    {
        stack.Pop();
    }

    public int Top()
    {
        return stack.Peek()[0];
    }

    public int GetMin()
    {
        return stack.Peek()[1];
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.Push(val);
 * obj.Pop();
 * int param_3 = obj.Top();
 * int param_4 = obj.GetMin();
 */